### PR TITLE
Undo old limited reader changes

### PR DESCRIPTION
**Pull request by harshilsharma63**
#### Summary
**Note- This PR is on top of #26148 and the twio will be merged together in master**

This PR removes the limited reader usage in individual in favour of global payload limits being added in this PR - https://github.com/mattermost/mattermost/pull/26148

This partially reverts the work of three PRs, removing the limited reader changes but keeping other changes-
1. [pr-mattermost-25579](https://github.com/mattermost/mattermost/pull/25579)
2. [pr-mattermost-25510](https://github.com/mattermost/mattermost/pull/25510)
3. [pr-mattermost-25943](https://github.com/mattermost/mattermost/pull/25943)

#### Release Note
```release-note
None.
```

### CODE REVIEW

**Review by larkox — APPROVED** [review]
LGTM

**Review by sbishel — APPROVED** [review]

File: server/public/model/utils.go [hunk 1]
```diff
@@ -487,39 +486,15 @@ func ArrayToJSON(objmap []string) string {
...
 
-// Deprecated: ArrayFromJSON is deprecated,
-// use SortedArrayFromJSON or NonSortedArrayFromJSON instead
 func ArrayFromJSON(data io.Reader) []string {
 	var objmap []string
+
 	json.NewDecoder(data).Decode(&objmap)
 	if objmap == nil {
 		return make([]string, 0)
 	}
```
Comment (sbishel) [inline server/public/model/utils.go:495]: Should we add the duplicate check that was added to the new functions. As I recall @agnivade was pretty insistent on this check.

File: server/public/model/utils.go [hunk 2]
```diff
@@ -555,16 +530,6 @@ func StringInterfaceFromJSON(data io.Reader) map[string]any {
 	return objmap
 }
 
-func StructFromJSONLimited[V any](data io.Reader, maxBytes int64, obj *V) error {
```
Comment (sbishel) [inline server/public/model/utils.go:558]: I would prefer we kept a `StructFromJSON` function that all the methods used. That way we do all our deserialization in a common place and way.

**Review by sbishel — COMMENTED** [review]
A couple changes we may want to keep.

File: server/public/model/utils.go (hunk 1 above)
Comment (agnivade) [reply server/public/model/utils.go:495]: Yeah, it'd be good to keep that. It's an orthogonal problem, but sometimes the DB doesn't have strong integrity checks and if the input is not checked properly, duplicate elements might get inserted.

Comment (harshilsharma63) [reply server/public/model/utils.go:495]: Done.

File: server/public/model/utils.go (hunk 2 above)
Comment (harshilsharma63) [reply server/public/model/utils.go:558]: Done.

File: server/channels/api4/channel.go [hunk 3]
```diff
@@ -425,18 +425,9 @@ func restoreChannel(c *Context, w http.ResponseWriter, r *http.Request) {
...
-		return
-	}
+	userIds := model.ArrayFromJSON(r.Body)
 	allowed := false
 
-	// single userId allowed if creating a self-channel
-	// NonSortedArrayFromJSON will remove duplicates, so need to add back
-	if len(userIds) == 1 && userIds[0] == c.AppContext.Session().UserId {
-		userIds = append(userIds, userIds[0])
-	}
```
Comment (esarafianou) [inline server/channels/api4/channel.go:439]: Why was this removed? The `ArrayFromJSON` calls `RemoveDuplicateStringsNonSort` so it also removes any duplicates like the current implementation does.

Comment (esarafianou) [inline server/channels/api4/channel.go:439]: Why was this removed? The ArrayFromJSON calls RemoveDuplicateStringsNonSort so it also removes any duplicates like the current implementation does.

File: server/channels/api4/channel.go [hunk 4]
```diff
@@ -1583,16 +1562,14 @@ func viewChannel(c *Context, w http.ResponseWriter, r *http.Request) {
...
 	c.RequireUserId()
 
-	channelIds, err := model.SortedArrayFromJSON(r.Body, *c.App.Config().ServiceSettings.MaximumPayloadSizeBytes)
-	if err != nil {
-		c.Err = model.NewAppError("readMultipleChannels", model.PayloadParseError, nil, "", http.StatusBadRequest).Wrap(err)
-		return
-	} else if len(channelIds) == 0 {
-		c.SetInvalidParam("channel_ids")
+	var channelIDs []string
+	err := json.NewDecoder(r.Body).Decode(&channelIDs)
```
Comment (esarafianou) [inline server/channels/api4/channel.go:1566]: Why are we using the json decoder here instead of the `ArrayFromJSON` ?

File: server/channels/api4/data_retention.go [hunk 5]
```diff
@@ -264,9 +264,10 @@ func searchTeamsInPolicy(c *Context, w http.ResponseWriter, r *http.Request) {
 func addTeamsToPolicy(c *Context, w http.ResponseWriter, r *http.Request) {
 	c.RequirePolicyId()
 	policyId := c.Params.PolicyId
-	teamIDs, err := model.SortedArrayFromJSON(r.Body, *c.App.Config().ServiceSettings.MaximumPayloadSizeBytes)
-	if err != nil {
-		c.Err = model.NewAppError("addTeamsToPolicy", model.PayloadParseError, nil, "", http.StatusBadRequest).Wrap(err)
+	var teamIDs []string
+	jsonErr := json.NewDecoder(r.Body).Decode(&teamIDs)
```
Comment (esarafianou) [inline server/channels/api4/data_retention.go:268]: Similar question. Why are we using the json decoder instead of the ArrayFromJSON? Is it due to the need for having them sorted? 

If there's a reason for the json decoder usages, should we call `model.RemoveDuplicateStrings` to remove any duplicates?

File: server/channels/api4/user.go [hunk 6]
```diff
@@ -955,15 +953,17 @@ func requireGroupAccess(c *web.Context, groupID string) *model.AppError {
...
 
 func getUsersByIds(c *Context, w http.ResponseWriter, r *http.Request) {
-	userIDs, err := model.SortedArrayFromJSON(r.Body, *c.App.Config().ServiceSettings.MaximumPayloadSizeBytes)
-	if err != nil {
-		c.Err = model.NewAppError("getUsersByIds", model.PayloadParseError, nil, "", http.StatusBadRequest).Wrap(err)
-		return
-	} else if len(userIDs) == 0 {
-		c.SetInvalidParam("user_ids")
+	var userIDs []string
+	err := json.NewDecoder(r.Body).Decode(&userIDs)
```
Comment (esarafianou) [inline server/channels/api4/user.go:957]: Why not use the ArrayFromJSON instead` that also does de-duplication? It it because of the need to keep the order?

File: server/channels/api4/user_local.go [hunk 7]
```diff
@@ -233,25 +233,27 @@ func localGetUsers(c *Context, w http.ResponseWriter, r *http.Request) {
...
+
+	if len(userIDs) == 0 {
 		c.SetInvalidParam("user_ids")
 		return
 	}
 
+	// we remove the duplicate IDs as it can bring a significant load to the
+	// database.
+	userIDs = model.RemoveDuplicateStrings(userIDs)
+
```
Comment (esarafianou) [inline server/channels/api4/user_local.go:246]: ArrayFromJSON already does a de-duplicatin. why calling `model.RemoveDuplicateStrings` again?

**Review by esarafianou — CHANGES_REQUESTED** [review]
@harshilsharma63 I omitted some comments that were similar to my existing ones, because my main question is the following:

Why is some cases we replace the `NonSortedArrayFromJSON` or `SortedArrayFromJSON` with the `ArrayFromJSON` and in some others use the json decoder directly?

**Comment by harshilsharma63** [comment]
> Why is some cases we replace the `NonSortedArrayFromJSON` or `SortedArrayFromJSON` with the `ArrayFromJSON` and in some others use the json decoder directly?

@larkox because they were added only to fix the side-effects of new limited reader functions added previously. Now that there are none of those limited reader functions, there are also none of their side-effects so no need of the special sorted and non-sorted methods. The PR undos the stuff in the list of PRs linked in the comment.

**Comment by esarafianou** [comment]
@harshilsharma63 The `ArrayToJson` though includes the de-duplication that both @sbishel and @agnivade mentioned that it's valuable.

**Comment by harshilsharma63** [comment]
> @harshilsharma63 The `ArrayToJson` though includes the de-duplication that both @sbishel and @agnivade mentioned that it's valuable.

Got it. I'll add it back.

**Comment by harshilsharma63** [comment]
Closing this in favour of #26263
